tracker/maintenance.py

```diff
@@ -0,0 +1,99 @@
+from tracker import db
+from tracker.model import CVE, CVEGroupEntry, CVEGroup, CVEGroupPackage, Package
+from tracker.model.enum import Status, affected_to_status, status_to_affected, highest_severity
+from tracker.pacman import search, get_configpath, archs, primary_arch
+from sqlalchemy import func
+from collections import defaultdict
+from subprocess import check_output
+from datetime import datetime
+
+
+def update_group_status():
+    groups = (db.session.query(CVEGroup, func.group_concat(CVEGroupPackage.pkgname, ' '))
+                .join(CVEGroupPackage)
+                .filter(CVEGroup.status.in_([Status.vulnerable, Status.testing]))
+                .group_by(CVEGroupPackage.group_id)).all()
+    for group, pkgnames in groups:
+        pkgnames = pkgnames.split(' ')
+        new_status = affected_to_status(status_to_affected(group.status), pkgnames[0], group.fixed)
+        if group.status is not new_status:
+            print('    --> Updated {}: {} -> {}'.format(group.name, group.status, new_status))
+        group.status = new_status
+    db.session.commit()
+
+
+def recalc_group_status():
+    groups = (db.session.query(CVEGroup, func.group_concat(CVEGroupPackage.pkgname, ' '))
+                .join(CVEGroupPackage)
+                .group_by(CVEGroupPackage.group_id)).all()
+    for group, pkgnames in groups:
+        pkgnames = pkgnames.split(' ')
+        new_status = affected_to_status(status_to_affected(group.status), pkgnames[0], group.fixed)
+        if group.status is not new_status:
+            print('    --> Updated {}: {} -> {}'.format(group.name, group.status, new_status))
+        group.status = new_status
+    db.session.commit()
+
+
+def recalc_group_severity():
+    entries = (db.session.query(CVEGroup, CVEGroupEntry, CVE)
+               .join(CVEGroupEntry).join(CVE)
+               .group_by(CVEGroupEntry.group_id).group_by(CVE.id)).all()
+    issues = defaultdict(set)
+    for group, entry, issue in entries:
+        issues[group].add(issue)
+    for group, issues in issues.items():
+        new_severity = highest_severity([issue.severity for issue in issues])
+        if group.severity is not new_severity:
+            print('    --> Updated {}: {} -> {}'.format(group.name, group.severity, new_severity))
+        group.severity = new_severity
+    db.session.commit()
+
+
+def update_package_cache():
+    print('    --> Querying alpm database...')
+    packages = search('', filter_duplicate_packages=False, sort_results=False)
+    pkgbases = {}
+    packages_by_arch = defaultdict(set)
+    # assign packages per arch list (unify any into 64bit)
+    for package in packages:
+        arch = package.arch if package.arch != 'any' else primary_arch
+        packages_by_arch[arch].add(package.name)
+
+    print('    --> Collecting pkgbases...')
+    # query all pkgbases per arch
+    for arch in archs:
+        cmd = ['expac', '-t', '%s', '-S', '--config', get_configpath(arch), '%n %e %v %b']
```

Comment:
Didn't we switch to alpm's pkgbase in master?

Reply:
nice catch, fixed